blockwork/workflows/workflow.py

```diff
@@ -40,18 +42,43 @@ class Workflow(RegisteredClass, metaclass=Singleton):
     PROJECT_TYPE = base.Project
     TARGET_TYPE = base.Element
 
-    def __init__(self, ctx: Context, site: base.Site, project: base.Project, target: base.Element) -> None:
+    def __init__(self, ctx: Context, project: str, target: str, **options) -> None:
         self.ctx = ctx
-        self.site = site
-        self.project = project
-        self.target = target
+
+        site_parser = parsers.Site(ctx)
+        site_path = ctx.site
+        self.site = site_parser(self.SITE_TYPE).parse(site_path)
+
+        project_parser = parsers.Project(ctx, self.site)
+        project_path = Path(self.site.projects[project])
+        self.project = project_parser(self.PROJECT_TYPE).parse(project_path)
+
+        target_parser = parsers.Element(ctx, self.site, self.project)
+        self.target = target_parser.parse_target(target, self.TARGET_TYPE)
+
+        self.workflow_options = options
+
+    @staticmethod
+    def register():
+        def inner(workflow: type["Workflow"]) -> "Workflow":
+            # This registers the workflow as a subcommand with default options
+            workflow = click.pass_obj(workflow)
+            wf_command = click.command(workflow)
+            wf_command = click.option('--project', '-p', type=str, required=True)(wf_command)
+            wf_command = click.option('--target', '-t', type=str, required=True)(wf_command)
+            wf.add_command(wf_command, name=workflow.__name__.lower())
+            return wf_command
+        return inner
 
     def depth_first_elements(self, element: base.Element) -> Iterable[base.Element]:
         'Recures elements and yields depths first'
         for sub_element in element.iter_elements():
             yield from self.depth_first_elements(sub_element)
         yield element
 
+    # Note this is run as a post-init hook so subclasses can add attributes
+    # after the init by overriding the method but before the run.
+    @InitHooks.post
```

Comment:
I'm sorry but I don't follow this - it seems like `InitHooks` may be a misnomer if it's wrapping around a `run` method? When in the program's lifecycle does this `run` get evaluated?

Reply:
This will be evaluated immediately after `__init__`.

Inithooks is used to ensure that a method is called before or after any subclass init.

Previously run was just called at the end of init, but the reason I used inithooks here is that I want the base class to process base arguments, but for subclasses to be able to add and process further arguments. If run() is called at the end of the base classes init then it's difficult to do this and manage super().__init__

The reason run needs to be called from init anyway is that this is a click command - which I don't think allows a two step approach nicely